app/src/main/java/org/p2p/wallet/common/EncryptedSharedPreferences.kt

```diff
@@ -0,0 +1,255 @@
+import timber.log.Timber
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty
+import org.p2p.core.crypto.Hex
+import org.p2p.wallet.common.crypto.keystore.DecodeCipher
+import org.p2p.wallet.common.crypto.keystore.EncodeCipher
+import org.p2p.wallet.common.crypto.keystore.KeyStoreWrapper
+
+class EncryptedSharedPreferences(
+    private val keyStoreWrapper: KeyStoreWrapper,
+    private val sharedPreferences: SharedPreferences,
+    private val gson: Gson
+) {
+    var loggingTag: String = "EncryptedSharedPreferences"
+
+    fun saveString(key: String, data: String) {
+        tryWithLog(key) {
+            val encodedData = keyStoreWrapper.encode(key, data)
+            sharedPreferences.edit { putString(key, encodedData) }
+        }
+    }
+
+    fun saveString(key: String, data: String, cipher: EncodeCipher) {
+        tryWithLog(key) {
+            val encodedData = keyStoreWrapper.encode(cipher, data)
+            sharedPreferences.edit { putString(key, encodedData) }
+        }
+    }
+
+    fun getString(key: String): String? {
+        return tryWithLog(key) {
+            val encodedData = sharedPreferences.getString(key, null)
+            encodedData?.let { keyStoreWrapper.decode(key, it) }
+        }
+    }
+
+    fun getString(key: String, cipher: DecodeCipher): String? {
+        return tryWithLog(key) {
+            val encodedData = sharedPreferences.getString(key, null)
+            encodedData?.let { keyStoreWrapper.decode(cipher, it) }
+        }
+    }
+
+    fun <T> saveObject(key: String, data: T) {
+        tryWithLog(key) {
+            val objectAsJson = gson.toJson(data)
+            val encodedData = keyStoreWrapper.encode(key, objectAsJson)
+            sharedPreferences.edit { putString(key, encodedData) }
+        }
+    }
+
+    /**
+     * !!! WARNING !!!
+     * DOESN'T WORK WITH CLASSES
+     */
+    @Deprecated(message = "DOESN'T WORK, use only with primitive types like String")
+    fun <T> saveObjectList(key: String, data: List<T>) {
+        saveObject(key, data)
+    }
+
+    @Throws(IllegalArgumentException::class)
+    fun <T : Any> getObject(key: String, type: KClass<T>): T? {
+        return tryWithLog(key) {
+            sharedPreferences.getString(key, null)
+                ?.let { keyStoreWrapper.decode(key, it) }
+                ?.let { gson.fromJson(it, type.java) }
+        }
+    }
+
+    /**
+     * !!! WARNING !!!
+     * DOESN'T WORK WITH CLASSES
+     */
+    @Deprecated(message = "DOESN'T WORK, use only with primitive types like String")
+    fun <T : Any> getObjectList(key: String): List<T> {
+        return tryWithLog(key) {
+            sharedPreferences.getString(key, null)
+                ?.let { keyStoreWrapper.decode(key, it) }
+                ?.let { gson.fromJson(it, object : TypeToken<List<T>>() {}.type) }
+                ?: emptyList()
+        }
+    }
+
+    fun saveBytes(key: String, data: ByteArray) {
```

Comment:
why `saveBytes` but `putBoolean`, `putLong` etc? I think the naming should be revised